File: sdk/typescript/src/modules/epoch.ts
import type { ChainClient } from "../infrastructure/chain-client.js";
import type {
  EpochInfoResponse,
  QualificationResponse,
  WindowActivity,
} from "../types/responses.js";
import type { SigningService } from "../infrastructure/signing-service.js";
import {
  computeWindow,
  formatProgress,
} from "../utils/epoch.js";

export class EpochModule {
  constructor(
    private readonly chainClient: ChainClient,
    private readonly signingService: SigningService,
  ) {}

  async getInfo(): Promise<EpochInfoResponse> {
    // Query proto response (4 fields)
    const proto = await this.chainClient.queryRest<{
      current_epoch: string;
      current_window: string;
      epoch_start_block: string;
      blocks_until_next_epoch: string;
    }>("/seocheon/activity/v1/epoch-info");

    const params = await this.getActivityParams();
    const windowLength = Math.floor(
      params.epoch_length / params.windows_per_epoch,
    );

    const block = await this.chainClient.getLatestBlock();
    const blockHeight = block.header.height;

    const currentEpoch = parseInt(proto.current_epoch, 10);
    const currentWindow = parseInt(proto.current_window, 10);
    const epochStartBlock = parseInt(proto.epoch_start_block, 10);
    const blocksUntilNextEpoch = parseInt(proto.blocks_until_next_epoch, 10);

    const epochEndBlockHeight = epochStartBlock + params.epoch_length - 1;
    const epochElapsed = blockHeight - epochStartBlock + 1;

    const winStartBlock = epochStartBlock + currentWindow * windowLength;
    const winEndBlock = winStartBlock + windowLength - 1;
    const windowElapsed = blockHeight - winStartBlock + 1;

    return {
      block_height: blockHeight,
      epoch_number: currentEpoch,
      epoch_start_block: epochStartBlock,
      epoch_end_block: epochEndBlockHeight,
      epoch_progress: formatProgress(epochElapsed, params.epoch_length),
      window_number: currentWindow,
      window_start_block: winStartBlock,
      window_end_block: winEndBlock,
      window_progress: formatProgress(windowElapsed, windowLength),
      blocks_until_next_window: winEndBlock - blockHeight,
      blocks_until_next_epoch: blocksUntilNextEpoch,
    };
  }

  async getQualification(
    nodeId?: string,
    epochNumber?: number,
  ): Promise<QualificationResponse> {
    const effectiveNodeId = nodeId ?? (await this.resolveOwnNodeId());

    // Get epoch info to determine current epoch
    const epochInfo = await this.chainClient.queryRest<{
      current_epoch: string;
      current_window: string;
      epoch_start_block: string;
      blocks_until_next_epoch: string;
    }>("/seocheon/activity/v1/epoch-info");

    const currentEpoch = parseInt(epochInfo.current_epoch, 10);
    const currentWindow = parseInt(epochInfo.current_window, 10);
    const effectiveEpoch = epochNumber ?? currentEpoch;

    const params = await this.getActivityParams();

    // Query node epoch activity
    const nodeEpoch = await this.chainClient.queryRest<{
      summary: {
        active_windows: string;
        total_activities: string;
        eligible: boolean;
      };
      quota_used: string;
      quota_limit: string;
    }>(
      `/seocheon/activity/v1/nodes/${effectiveNodeId}/epochs/${effectiveEpoch}`,
    );

    const activeWindows = parseInt(nodeEpoch.summary.active_windows, 10);

    // Calculate elapsed windows
    const elapsedWindows =
      effectiveEpoch === currentEpoch
        ? currentWindow + 1
        : params.windows_per_epoch;

    const remainingNeeded = Math.max(
      0,
      params.min_active_windows - activeWindows,
    );
    const remainingWindows = params.windows_per_epoch - elapsedWindows;
    const canStillQualify =
      activeWindows + remainingWindows >= params.min_active_windows;

    // Build window detail from activities
    const activitiesResponse = await this.chainClient.queryRest<{
      activities: Array<{
        activity_hash: string;
        block_height: string;
      }>;
    }>(
      `/seocheon/activity/v1/nodes/${effectiveNodeId}/activities?epoch=${effectiveEpoch}`,
    );

    const windowCounts = new Map<number, number>();
    for (const record of activitiesResponse.activities ?? []) {
      const wn = computeWindow(
        parseInt(record.block_height, 10),
        params.epoch_length,
        params.windows_per_epoch,
      );
      windowCounts.set(wn, (windowCounts.get(wn) ?? 0) + 1);
    }

    const windowDetail: WindowActivity[] = [];
    for (let w = 0; w < params.windows_per_epoch; w++) {
      const count = windowCounts.get(w) ?? 0;
      windowDetail.push({
        window_number: w,
        submission_count: count,
        has_activity: count > 0,
      });
    }

    return {
      epoch_number: effectiveEpoch,
      total_windows: params.windows_per_epoch,
      elapsed_windows: elapsedWindows,
      active_windows: activeWindows,
      required_windows: params.min_active_windows,
      is_qualified: nodeEpoch.summary.eligible,
      remaining_needed: remainingNeeded,
      can_still_qualify: canStillQualify,
      window_detail: windowDetail,
    };
  }

  private async resolveOwnNodeId(): Promise<string> {
    const address = await this.signingService.getAddress();
    const response = await this.chainClient.queryRest<{
      node: { id: string };
    }>(`/seocheon/node/v1/nodes/by-agent/${address}`);
    return response.node.id;
  }

  private async getActivityParams(): Promise<{
    epoch_length: number;
    windows_per_epoch: number;
    min_active_windows: number;
  }> {
    const response = await this.chainClient.queryRest<{
      params: {
        epoch_length: string;
        windows_per_epoch: string;
        min_active_windows: string;
      };
    }>("/seocheon/activity/v1/params");
    return {
      epoch_length: parseInt(response.params.epoch_length, 10),
      windows_per_epoch: parseInt(response.params.windows_per_epoch, 10),
      min_active_windows: parseInt(response.params.min_active_windows, 10),
    };
  }
}

File: sdk/python/seocheon/modules/epoch.py
"""Epoch module for the Seocheon SDK."""

from __future__ import annotations

from seocheon.constants.chain import EPOCH_LENGTH, MIN_ACTIVE_WINDOWS, WINDOWS_PER_EPOCH
from seocheon.errors.errors import ErrQueryFailed
from seocheon.infrastructure.chain_client import ChainClient
from seocheon.types.responses import (
    EpochInfoResponse,
    QualificationResponse,
    WindowActivity,
)


class EpochModule:
    """Provides epoch-related operations."""

    def __init__(self, client: ChainClient) -> None:
        self._client = client

    async def get_info(self) -> EpochInfoResponse:
        """Return the current epoch and window state."""
        try:
            data = await self._client.query_rest("/seocheon/activity/v1/epoch-info")
        except Exception as e:
            raise ErrQueryFailed from e

        params = await self._get_params()
        block = await self._client.get_latest_block()

        current_epoch = int(data.get("current_epoch", 0))
        current_window = int(data.get("current_window", 0))
        epoch_start_block = int(data.get("epoch_start_block", 0))
        blocks_until_next_epoch = int(data.get("blocks_until_next_epoch", 0))

        window_length = params["epoch_length"] // params["windows_per_epoch"]
        epoch_end_block = epoch_start_block + params["epoch_length"] - 1
        window_start_block = epoch_start_block + (current_window * window_length)
        window_end_block = window_start_block + window_length - 1

        epoch_elapsed = block["height"] - epoch_start_block + 1
        window_elapsed = block["height"] - window_start_block + 1

        return EpochInfoResponse(
            block_height=block["height"],
            epoch_number=current_epoch,
            epoch_start_block=epoch_start_block,
            epoch_end_block=epoch_end_block,
            epoch_progress=f"{epoch_elapsed}/{params['epoch_length']}",
            window_number=current_window,
            window_start_block=window_start_block,
            window_end_block=window_end_block,
            window_progress=f"{window_elapsed}/{window_length}",
            blocks_until_next_window=window_end_block - block["height"],
            blocks_until_next_epoch=blocks_until_next_epoch,
        )

    async def get_qualification(self, node_id: str, epoch_number: int = 0) -> QualificationResponse:
        """Return the activity reward qualification status for a node."""
        if not node_id:
            raise ValueError("node_id is required for qualification query")

        effective_epoch = epoch_number
        if effective_epoch == 0:
            info = await self.get_info()
            effective_epoch = info.epoch_number

        path = f"/seocheon/activity/v1/nodes/{node_id}/epochs/{effective_epoch}"
        try:
            data = await self._client.query_rest(path)
        except Exception as e:
            raise ErrQueryFailed from e

        params = await self._get_params()
        summary = data.get("summary", {})
        active_windows = int(summary.get("active_windows", 0))

        epoch_info = await self.get_info()
        if effective_epoch == epoch_info.epoch_number:
            elapsed_windows = epoch_info.window_number + 1
        else:
            elapsed_windows = params["windows_per_epoch"]

        remaining_needed = max(0, params["min_active_windows"] - active_windows)
        remaining_windows = params["windows_per_epoch"] - elapsed_windows
        can_still_qualify = (active_windows + remaining_windows) >= params["min_active_windows"]

        # Build window detail
        window_detail = [
            WindowActivity(window_number=w, submission_count=0, has_activity=False)
            for w in range(params["windows_per_epoch"])
        ]

        # Enrich from activities query
        try:
            act_path = f"/seocheon/activity/v1/nodes/{node_id}/activities?epoch={effective_epoch}"
            act_data = await self._client.query_rest(act_path)
            window_length = params["epoch_length"] // params["windows_per_epoch"]
            for act in act_data.get("activities", []):
                h = int(act.get("block_height", 0))
                wn = ((h - 1) % params["epoch_length"]) // window_length
                if 0 <= wn < params["windows_per_epoch"]:
                    window_detail[wn].submission_count += 1
                    window_detail[wn].has_activity = True
        except Exception:
            pass

        return QualificationResponse(
            epoch_number=effective_epoch,
            total_windows=params["windows_per_epoch"],
            elapsed_windows=elapsed_windows,
            active_windows=active_windows,
            required_windows=params["min_active_windows"],
            is_qualified=summary.get("eligible", False),
            remaining_needed=remaining_needed,
            can_still_qualify=can_still_qualify,
            window_detail=window_detail,
        )

    async def _get_params(self) -> dict:
        try:
            data = await self._client.query_rest("/seocheon/activity/v1/params")
            params = data.get("params", {})
            el = int(params.get("epoch_length", 0)) or EPOCH_LENGTH
            wpe = int(params.get("windows_per_epoch", 0)) or WINDOWS_PER_EPOCH
            maw = int(params.get("min_active_windows", 0)) or MIN_ACTIVE_WINDOWS
            return {"epoch_length": el, "windows_per_epoch": wpe, "min_active_windows": maw}
        except Exception:
            return {"epoch_length": EPOCH_LENGTH, "windows_per_epoch": WINDOWS_PER_EPOCH, "min_active_windows": MIN_ACTIVE_WINDOWS}
